src/renderers/shared/reconciler/ReactCompositeComponent.js

```diff
@@ -659,16 +660,16 @@ var ReactCompositeComponentMixin = {
         transaction,
         this._context
       );
-    }
-
-    if (this._pendingStateQueue !== null || this._pendingForceUpdate) {
+    } else if (this._pendingStateQueue !== null || this._pendingForceUpdate) {
       this.updateComponent(
         transaction,
         this._currentElement,
         this._currentElement,
         this._context,
         this._context
       );
+    } else {
+      this._updateBatchNumber = null;
```

Comment:
Yeah. :( setState calls enqueueSetState and enqueueCallback which each create a separate batch. That is almost certainly unintended but I left it as-is for now.